icu4c/source/i18n/calendar.cpp

```diff
@@ -733,10 +733,10 @@ fTime(0),
 fLenient(true),
 fZone(nullptr),
 fRepeatedWallTime(UCAL_WALLTIME_LAST),
-fSkippedWallTime(UCAL_WALLTIME_LAST)
+fSkippedWallTime(UCAL_WALLTIME_LAST),
+validLocale(new CharString()),
+actualLocale(new CharString())
 {
-    validLocale[0] = 0;
-    actualLocale[0] = 0;
     if (U_FAILURE(success)) {
         delete zone;
         return;
```

Comment:
2 lines below: check if the two locale pointers are nullptr; if so: mem-alloc-error, and delete the zone if you didn't use a LocalPointer for it